### PR TITLE
Add separate beatmap update flow to handle edge cases better

### LINKED ISSUE #19369: Beatmap with missing online ID duplicating when updated
### Discussed in https://github.com/ppy/osu/discussions/19342

<div type='discussions-op-text'>

<sup>Originally posted by **iSimpFor** July 24, 2022</sup>
mapset in question: https://osu.ppy.sh/beatmapsets/17819

when updating the mentioned beatmap, the mapset duplicates itself on the carousel while also failing to actually update  

[updater.log](https://github.com/ppy/osu/files/9174894/updater.log)
[runtime.log](https://github.com/ppy/osu/files/9174895/runtime.log)
[performance-update.log](https://github.com/ppy/osu/files/9174896/performance-update.log)
[performance-input.log](https://github.com/ppy/osu/files/9174897/performance-input.log)
[performance-draw.log](https://github.com/ppy/osu/files/9174898/performance-draw.log)
[performance-audio.log](https://github.com/ppy/osu/files/9174899/performance-audio.log)
[performance.log](https://github.com/ppy/osu/files/9174900/performance.log)
[network.log](https://github.com/ppy/osu/files/9174901/network.log)
[database.log](https://github.com/ppy/osu/files/9174902/database.log)

picture of this happening:
osu_2022-07-23_19-45-32

version: 2022.723.0</div>

### PR DESCRIPTION
**Pull request by peppy**
- Closes #19369 (b7f6413bce)
- Fixes loss of scores when updating a beatmap (on beatmaps which weren't touched) (912218e123)
- Fixes date added being refreshed unexpectedly during updates (2363a3fb7b)
- Clean up old versions of beatmap better (leaving soft deleted versions of only beatmaps which had local modifications) (e5ad07454c)

Pushing this to allow any feedback on the general flow. Will add more tests over the next day.

- [x] Add more test coverage

A few standout commits which aren't mentioned above:

### 2e14d8730c Move implementation of updating a beatmap to `BeatmapImporter`

Required to be able to write isolated tests without constructing a `BeatmapManager` or `BeatmapUpdater`. But also feels like it's probably the best place to put this logic (if not `BeatmapManager`, I guess).

### 6a3e8e31de Centralise calls to reset online info of a `BeatmapInfo`

Until now we were only resetting the `OnlineID` in multiple places, which leaves a beatmap in a state where it *thinks* it can still receive updates, even though the intention is that it shouldn't. This can be split out into its own PR if required.

### CODE REVIEW

File: osu.Game.Tests/Database/BeatmapImporterTests.cs [hunk 1]
```diff
@@ -670,6 +670,116 @@ public void TestImportThenDeleteThenImportOptimisedPath()
             });
         }
 
+        [Test]
```
Comment (peppy) [inline osu.Game.Tests/Database/BeatmapImporterTests.cs:673]: I wanted to put these in a separate file, but splitting out this test feels like it should be done as a separate pass. There's a lot of helper methods and reorganisation to be done.

**Comment by smoogipoo** [comment]
I can't make sense of this PR and the classes contained.

- Several classes take a `BeatmapManager beatmapImporter` parameter: https://github.com/ppy/osu/pull/19378/files#diff-29aba4e3a33389232bd558484671c27bb77c3c3e1f6849e3fddf270c7d1b197aL240-R240 https://github.com/ppy/osu/pull/19378/files#diff-778d37b33149c32bd0d3e922e7aee595529feed842e7402d8cced78f6e46a476L112-R111. Why are these not `BeatmapImporter`.
- `BeatmapManager` has many `Import*` methods that forward _to_ `BeatmapImporter` which indicates that `BeatmapImporter` _should_ be the main receiver of all of these parameters https://github.com/ppy/osu/pull/19378/files#diff-3af5d6a981ffc459f672238cbabbeebc1edf263a903539ceb6f3ada58894a991R411-R412.
- `BeatmapModelDownloader` now has an `Import` and `Update` method. Yes, `Import` is protected but that doesn't make it easy for understanding the structure of things. https://github.com/ppy/osu/pull/19378/files#diff-92c05e56f1c8835bd73188256c2e3d35ee1d650fc49ace5809f5329776a2ee95R29-R37
- Similar to the above `ModelDownloader` also has an `Import` method https://github.com/ppy/osu/pull/19378/files#diff-5b9125d235ca287e338a8abe10f5e9841d7b92157c73a1eb06620d3a7d5cc8cbR108-R118
- `beatmapDownloader.Download(x)` -> `beatmapDownloader.Update(x)` looks weird https://github.com/ppy/osu/pull/19378/files#diff-c7cbb43ab53ebfe6453def12cb27ca53cfbf54e94702ee0c852c681ba0622ffdL93-R93
-

File: osu.Game/Beatmaps/Drawables/BundledBeatmapDownloader.cs [hunk 2]
```diff
@@ -109,7 +108,7 @@ private void queueDownloads(string[] sourceFilenames, int? limit = null)
 
         private class BundledBeatmapModelDownloader : BeatmapModelDownloader
         {
-            public BundledBeatmapModelDownloader(IModelImporter<BeatmapSetInfo> beatmapImporter, IAPIProvider api)
+            public BundledBeatmapModelDownloader(BeatmapManager beatmapImporter, IAPIProvider api)
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/Drawables/BundledBeatmapDownloader.cs:111]: Can this be reverted?

**Comment by peppy** [comment]
I've applied changes in 8370ca9 to address the issues you've mentioned (mostly). We agree that there needs to be better separation of concerns (ie. `ModelDownloader` should probably not be importing in the first place), but that will happen as a follow-up.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Beatmaps/Drawables/BundledBeatmapDownloader.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Beatmaps/Drawables/BundledBeatmapDownloader.cs]: Yeah, missed it 😓

**Comment by smoogipoo** [comment]
Updated structure is much more readable :+1:

**Comment by peppy** [comment]
Tests added, all ready to go.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**